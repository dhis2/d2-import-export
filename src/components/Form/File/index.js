import React from 'react'
import { SvgIcon } from 'material-ui'
import { FormControl, FormLabel } from '../material-ui'
import s from './styles.css'
import i18n from '@dhis2/d2-i18n'

function FileUploadIcon() {
    return (
        <SvgIcon>
            <path d="M0 0h24v24H0z" fill="none" />
            <path d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z" />
        </SvgIcon>
    )
}

export default class FileField extends React.Component {
    onClick = () => {
        this.fileRef && this.fileRef.click()
    }

    onChange = () => this.props.onChange(this.props.name, this.fileRef.files[0])

    render() {
        const { selected } = this.props
        let label = this.props.label
        if (selected) {
            label = selected.name
        }

        return (
            <FormControl className={s.formControl} onClick={this.onClick}>
                <input
                    type="file"
                    onChange={this.onChange}
                    ref={c => (this.fileRef = c)}
                    className={s.hiddenFileInput}
                />
                <FileUploadIcon className={s.button} />
                <FormLabel className={s.formLabel}>
                    {label || i18n.t('Choose a file to upload')}
                </FormLabel>
            </FormControl>
        )
    }
}
