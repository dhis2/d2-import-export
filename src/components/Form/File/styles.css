form .formControl {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: flex-end;
    cursor: pointer;
    width: 100%;
    margin: 15px 0 5px 0;
}

form .formLabel {
    margin-left: 10px;
    margin-bottom: 5px;
    cursor: pointer;
    user-select: none;
    color: #000;
    font-size: 13px;
    font-weight: 500;
    text-transform: uppercase;
}
.button {
    margin-top: 10px;
}
.hiddenFileInput {
    display: none;
}
