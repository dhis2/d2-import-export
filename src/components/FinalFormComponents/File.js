import { useField } from 'react-final-form'
import { Help } from '@dhis2/ui-core'
import React, { useRef, useCallback } from 'react'
import propTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import styles from './File.module.css'

const FileUploadIcon = () => (
    <svg className={styles.svg}>
        <path d="M0 0h24v24H0z" fill="none" />
        <path d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z" />
    </svg>
)

export const File = ({ name, dataTest }) => {
    const ref = useRef()
    const { input, meta } = useField(name, {
        type: 'file',
        parse: () => ref.current.files[0],
        validate: value => (!!value ? undefined : i18n.t('Required')),
    })
    const onClick = useCallback(() => {
        ref.current.click()
    }, [ref])

    const { value, ...withoutValue } = input
    const label = value ? value.name : i18n.t('Choose a file to upload') + ' *'

    return (
        <div className={styles.container} data-test={dataTest}>
            <input {...withoutValue} ref={ref} className={styles.input} />

            <button type="button" onClick={onClick} className={styles.button}>
                <FileUploadIcon />
                <span className={styles.label}>{label}</span>
            </button>

            {meta.touched && meta.error && (
                <div className={styles.error}>
                    <Help error>{meta.error}</Help>
                </div>
            )}
        </div>
    )
}

File.propTypes = {
    name: propTypes.string,
}
