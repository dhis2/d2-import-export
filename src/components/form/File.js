import { useField } from 'react-final-form'
import React, { Fragment, useRef, useCallback } from 'react'
import propTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import styles from './File.module.css'

const FileUploadIcon = () => (
    <svg className={styles.svg}>
        <path d="M0 0h24v24H0z" fill="none" />
        <path d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z" />
    </svg>
)

export const File = ({ name }) => {
    const ref = useRef()
    const { input, meta } = useField(name, {
        type: 'file',
        parse: () => ref.current.files[0],
    })
    const onClick = useCallback(() => {
        ref.current.click()
    }, [ref])

    const { value, ...withoutValue } = input
    const label = value ? value.name : i18n.t('Choose a file to upload')

    return (
        <Fragment>
            <input {...withoutValue} ref={ref} className={styles.input} />

            <button onClick={onClick} className={styles.container}>
                <FileUploadIcon />
                <span className={styles.label}>{label}</span>
            </button>
        </Fragment>
    )
}

File.propTypes = {
    name: propTypes.string,
}
