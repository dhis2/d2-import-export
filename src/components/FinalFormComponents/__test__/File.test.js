import React from 'react'
import renderer from 'react-test-renderer'
import { File } from '../File'

jest.mock('react-final-form', () => ({
    useField: jest.fn(() => ({
        input: {
            name: 'Name',
            value: '',
            onChange: jest.fn(),
        },
        meta: {
            touched: false,
            error: '',
        },
    })),
}))

describe('Form component - File', () => {
    it('should render correctly', () => {
        const file = renderer.create(<File name="upload" />).toJSON()

        expect(file).toMatchSnapshot()
    })
})
