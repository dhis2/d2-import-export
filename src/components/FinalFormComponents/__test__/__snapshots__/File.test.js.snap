// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component - File should render correctly 1`] = `
<div
  className="container"
>
  <input
    className="input"
    name="Name"
    onChange={[MockFunction]}
  />
  <button
    className="button"
    onClick={[Function]}
  >
    <svg
      className="svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
      />
    </svg>
    <span
      className="label"
    >
      Choose a file to upload *
    </span>
  </button>
</div>
`;
